Copy edits on JS3S1 (#432)
CodeYourFuture/curriculum: content/js3/blocks/simplifying-element-creation/index.md

+++
title = 'Simplifying element creation'
headless = true
time = 45
facilitation = false
emoji= '🧩'
[objectives]
    1='Extract functions for common tasks'
    2='Use template tags to simplify element initialisation'
+++

We now have a card showing all of the information for one film.

The code we have is quite repetitive and verbose. It does similar things lots of times.

Let's look at two ways we could simplify this code.

> ### Refactoring: Extraction a function

Reviewer: ```suggestion
### Refactoring: Extracting a function
```

One way we can simplify this code is to refactor it.

{{<note type="tip" title="Definition: refactoring">}}

> The process of updating our code quality (without changing the implementation) is called **refactoring**.

Reviewer: ```suggestion
To refactor means to update our code _quality_ without changing the _implementation_.
```

    element.textContent = textContent;
    parentElement.append(element);
    return element;
}
```

And then rewrite our code to create the card like this:

```js
const film = {
  title: "Killing of Flower Moon",
  director: "Martin Scorsese",
  times: ["15:35"],
  certificate: "15",
  duration: 112,
};

function createChildElement(parentElement, tagName, textContent) {
    const element = document.createElement(tagName);
    element.textContent = textContent;
    parentElement.append(element);
    return element;
}

const card = document.createElement("section");

createChildElement(card, "h3", film.title);

createChildElement(card, "p", `Director: ${film.director}`);

const times = createChildElement(card, "ul", "");

for (const time of film.times) {
  createChildElement(times, "li", time);
}

createChildElement(card, "p", `${film.duration} minutes`);

createChildElement(card, "p", `Certificate: ${film.certificate}`);

document.body.append(card);
```

This code does exactly the same thing as the code we had before. By introducing a function we have introduced some advantages:
1. Our code is smaller, which can make it easier to read and understand what it's doing.

> 2. If we want to change how we create elements (e.g. if we wanted to also add a class attribute for each element), we only need to write the new code one time, not for every element.

Reviewer: ```suggestion
2. If we want to change how we create elements we only need to write the new code one time, not for every element. We could add a class attribute for each element easily.
```

> 3. We can see that each of the elements is being created the same way. Before, if we wanted to make sure the director element and duration element were created the same, we would have to compare several lines of code. Because we can see they're calling the same function, we know they're made the same way.

Reviewer: ```suggestion
3. 3. We can see that each element is being created the same way. Before, we would have to compare several lines of code to see this. Because we can see they're calling the same function, we know they're made the same way.
```

There are also some drawbacks to our refactoring:
1. If we want to change how we create some elements (e.g. we want to include an image of the director, replace the certificate text with a symbol, etc), we may have made it _harder_ to make these changes.
2. If we want to see how one particlar thing, like the director, is show, we need to look in a few places. This is something you will need to get used to, so it's good to start practising now!

### Using `<template>` tags

> A different way we could simplify this code is to use a different technique for creating elements.

Reviewer: ```suggestion
We could simplify this code with a different technique for creating elements.
```

Until now, we have only seen one way to create elements: `document.createElement`.

The DOM has another way of creating elements - we can _copy_ existing elements and then change them.

HTML has a useful tag designed to help make this easy, [the `<template>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template). When you add a `<template>` element to a page, it doesn't get displayed when the page loads.

> We can copy any DOM element, not just `<template>` tags. For this problem, we will uus a `<template>` tag.

Reviewer: ```suggestion
We can copy any DOM element, not just `<template>` tags. For this problem, we will use a `<template>` tag because it has some special powers.
```

HTML has a useful tag designed to help make this easy, [the `<template>` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template). When you add a `<template>` element to a page, it doesn't get displayed when the page loads.

> We can copy any DOM element, not just `<template>` tags. For this problem, we will use a `<template>` tag because it isn't displayed, but can be used to make elements which are.

Reviewer: ```suggestion
We can copy any DOM element, not just `<template>` tags. For this problem, we will use a `<template>` tag. 
```
